### PR TITLE
Phase angle

### PR DESCRIPTION
**Pull request by yihengwuKP**
# Brief Description
Implementation of cv: phase angle of monocyclic ring molecules, using [Cremer-Pople method](https://pubs.acs.org/doi/10.1021/ja00839a011). Notice that for ring with more than 5 members, there are (N-1)/2-1 phase angles defined in Cremer-Pople method, and this class only calculate the first one (m=2 in Eq 12 and Eq 13). Potentially m can be added as an argument but I haven't figured out how, should I add an `__init__` for the class `PhaseAngle`? I would appreciate some help on that! However, this is already enough for my usage. 

# Test
I've tested this implementation against the implementation in [`MDAnalysis.nuclinfo.phase_cp`](https://docs.mdanalysis.org/1.0.1/documentation_pages/analysis/nuclinfo.html). I found that it's tricky actually to install MDAnalysis into the pysages conda environment in midway3, as the MDAnalysis would install one version of scipy that messed up with jax. So I used another conda environment for the analysis.

## Code for testing
```python
#!/usr/bin/env python

import openmm as mm
import openmm.app as app
from openmm import unit
from math import ceil
import os
import numpy as np
from pysages.colvars.angles import PhaseAngle, DihedralAngle
from pysages.methods import HistogramLogger, Unbiased
import pysages

step_size = 1 * unit.femtosecond
nsteps = 1000
print("nsteps = {}".format(nsteps))

pdb = app.PDBFile('guanosine.fixed.pdb')
sugar_indices = {}
for i, atom in enumerate(pdb.topology.atoms()):
    if atom.name in ["O4'", "C4'", "C3'", "C2'", "C1'"]:
        sugar_indices[atom.name] = atom.index

print(f"dictionary of sugar indices: {sugar_indices}")


def generate_simulation():

    forcefield = app.ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
    system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.PME,
                                     nonbondedCutoff=1.2*unit.nanometer,
                                     constraints=app.HBonds)
    system.addForce(mm.MonteCarloBarostat(1*unit.bar,
                                          298*unit.kelvin,
                                          1000))

    integrator = mm.LangevinIntegrator(298*unit.kelvin, 5/unit.picosecond,
                                       step_size.in_units_of(unit.picosecond))
    simulation = app.Simulation(pdb.topology, system, integrator)
    simulation.context.setPositions(pdb.positions)

    print("minimizing energy...")
    simulation.minimizeEnergy()

    print("Using {} platform".format(simulation.context.getPlatform().getName()))
    simulation.reporters.append(app.DCDReporter('output.dcd', 1))

    return simulation


cvs = [PhaseAngle(
        [sugar_indices[nm] for nm in ["O4'", "C1'", "C2'", "C3'", "C4'"]])]
# notice that the order of the indices do matter for the calculation of phase angle
method = Unbiased(cvs)
callback = HistogramLogger(1)

raw_result = pysages.run(method, generate_simulation, nsteps, callback)
np.savetxt('phase_angle.txt', raw_result.callbacks[0].data)
```
the input and output files and the scripts: [phase_angle_example.zip](https://github.com/SSAGESLabs/PySAGES/files/9788453/phase_angle_example.zip)

Comparing the results:
```python

import numpy as np
import MDAnalysis as mda
import matplotlib.pyplot as plt
from MDAnalysis.analysis import nuclinfo
from math import sin, cos, atan2, pi, sqrt

myP = np.loadtxt("phase_angle.txt")

u = mda.Universe("guanosine.fixed.pdb", "output.dcd")
P = []
for t in u.trajectory:
    P.append(nuclinfo.phase_cp(u, 'A', 1))

myP=(myP+np.pi/2)*180/np.pi%360
plt.plot(myP[1:], 'o', label='pysages', alpha=.5)
plt.plot(np.array(P[:-1]), label='MDAnalysis')
plt.legend()
plt.show()


plt.figure(figsize=(5,5))
plt.scatter(myP[1:], P[:-1])
plt.plot([320, 350], [320, 350], color='red')
```
the result: 
image
image

There is this manual hack of myP[1:], P[:-1], because I found the logger result from histogromLogger seems to be shifted, related to issue #225

### CODE REVIEW

**Comment by yihengwuKP** [comment]
I think this failed CI/doc-test is related with issue #226

**Comment by yihengwuKP** [comment]
Ok, I've added another commit to fix the spelling issue!

File: docs/source/pysages_wordlist.txt
Comment (pabloferz) [inline docs/source/pysages_wordlist.txt:53]:
```suggestion
membered
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:134]:
```suggestion
    Notice that for ring with more than 5 atoms, there are (N-1)/2-1 phase angles, and
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:135]:
```suggestion
    this class only calculates the first one (m=2 in Equation 12 and Equation 13, or see
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:141]:
```suggestion
    (Also, notice that in this implementation of MDAnalysis, a 90 degree is added to the resulting
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:142]:
```suggestion
    phase angle by CP method, because it is convention to add 90 degrees to convert to the
```

**Comment by pabloferz** [comment]
Yes, you should be able to commit the suggestions, but you can also manually make the changes on your working branch and push them here.

File: pysages/colvars/angles.py
Comment (yihengwuKP) [reply pysages/colvars/angles.py:141]: ah, this makes sense to me!

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:137]:
```suggestion
    Also, the phase angle obtained via the Cremer-Pople method can be converted to the
```

File: pysages/colvars/angles.py
Comment (pabloferz) [inline pysages/colvars/angles.py:139]:
```suggestion
    [D. Cremer and J. A. Pople, JACS, 1974](https://pubs.acs.org/doi/10.1021/ja00839a011)
    Equations 4-14.
    Notice that for rings with more than 5 atoms, there are `(N - 1) / 2 - 1`
    phase angles, and this class only calculates the first one (`m = 2` in
    Equations 12 and 13, or see `pysages.colvars.angles.phase_angle` for math).
    Also, the phase angle obtained via the Cremer-Pople method can be converted
    to the Altona-Sundaralingam order parameter by adding `pi / 2` to the result
    and then converting to degrees.
```

File: pysages/colvars/angles.py
Comment (pabloferz) [inline pysages/colvars/angles.py:170]:
```suggestion
    :math:`P = \atan(b / a)`
```

**Review by pabloferz — DISMISSED** [review]

**Comment by yihengwuKP** [comment]
Oh, it seems I somehow dismissed Pablo's stale review? So I shouldn't have pushed another commit? I just implement another relavant cv, that's the square root of the a and b which are calculated during the calculation of the phase angle. Not sure if this implementation is the most efficient way of doing it though. This c.v. is the amplitude of the ring pucker and less useful than the phase angle, but might be good to have.

**Comment by InnocentBug** [comment]
No, this is not an issue. 
you automatically dismiss the review when you push changes. 
It just means that we have to review the changed version before we can merge. 

It doesn’t mean that you disregard his recommendations. More the opposite. 
when you are done with changes, just rerequest the review.

**Comment by yihengwuKP** [comment]
Thanks! I've added the doc string and it's now ready for review

**Comment by pabloferz** [comment]
After looking at the two CVs, I think there's a lot of room for simplifying the implementation. We should probably provide three classes here:

 1. A class called `RingPuckeringCoordinates`, this would be a multicomponent CV (see https://github.com/SSAGESLabs/PySAGES/blob/main/pysages/colvars/coordinates.py#L113)
 2. A class `RingPhaseAngle` whose `function` should call `ring_pluckering_coordinates`,
 3. And a class `RingAmplitude`, again, with `ring_amplitude` calling `ring_pluckering_coordinates`.

This should good enough for now even if we do not provide yet a way to compute the coordinates for each value of `m` in the Cremer-Pople method.

**Comment by yihengwuKP** [comment]
Ah, thanks for the guidance! This is what I don't know, and I'll try to implement it.

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:134]:
```suggestion
    So if the ring contains more than six atoms, there is more than one phase angle.
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:207]:
```suggestion
    So if the ring contains more than five atoms, there is more than one amplitude.
```

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:211]:
```suggestion
    to the Altona-Sundaralingam order parameter by multiplying the result by `102.5`.
```

**Comment by yihengwuKP** [comment]
Ok, I've implemented the multicomponent cv RingPuckeringCoordinates like @pabloferz suggested. I've tested them against MDAnalysis calculations and results are following:
image
image
image
image

**Comment by InnocentBug** [comment]
Ok, that looks great.
Awesome that you have tests for your code.
I know that you said, you have troubles with analysis and running the code.

But can you try to sort your test code into the `examples/openmm` directory.
And after that, please add your example to the automated testing in `.github/workflows/docker-ci.yml`.
You can find many examples there that make use of example scripts.
It might be that you have to prune the example a little bit in order to complete the simulation in about 15min on a single cpu.

It is OK if that is too short for actual physical testing, but we want to run your code contribution to make sure that it works fine.

Thank you.

**Comment by yihengwuKP** [comment]
cool, right now the example runs a few seconds in gpu. I'll try to include it in the example and the CI. I'll also fix the doc-test issues.

**Comment by yihengwuKP** [comment]
I didn't add it to `examples` because I thought they are all for the methods, should I just add a `openmm/unbiased` example and put my scripts there?

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:149]: 1025 or 102.5?

**Comment by InnocentBug** [comment]
Oh, I see.
I think the reason all examples are methods, is because that was the main focus of development.
I would include the example as `openmm/pucker`, so the CV instead of the method.

**Comment by yihengwuKP** [comment]
Cool, thanks for all your help! I'll do that

File: pysages/colvars/angles.py
Comment (ricalessandri) [reply pysages/colvars/angles.py:149]: ok!

File: pysages/colvars/angles.py
Comment (ricalessandri) [inline pysages/colvars/angles.py:150]:
```suggestion
```

File: examples/openmm/cv_pucker/unbiased.py
Comment (ricalessandri) [inline examples/openmm/cv_pucker/unbiased.py:57]: This part could probably go into a `def main` function, like, e.g., [here](https://github.com/SSAGESLabs/PySAGES/blob/main/examples/openmm/spectral_abf/alanine-dipeptide.py)

File: pysages/colvars/angles.py
Comment (yihengwuKP) [reply pysages/colvars/angles.py]: cool, that's much clearer!

File: examples/openmm/cv_pucker/unbiased.py
Comment (yihengwuKP) [reply examples/openmm/cv_pucker/unbiased.py:57]: ah, right, totally! I was a bit lazy

Comment (yihengwuKP) [reply examples/openmm/cv_pucker/unbiased.py:57]: I'll change it

File: .github/workflows/docker-ci.yml
Comment (InnocentBug) [inline .github/workflows/docker-ci.yml:162]: Is there a way to pre-record the expected CV and compare with the PySAGES result.

Comment (yihengwuKP) [reply .github/workflows/docker-ci.yml:162]: It's hard to make it purely deterministic. However, I have scripts to do post-hoc calculation using MDAnalysis and can compare. I guess I can change the unbiased.py, like adding an assert on the deviation between two results?

Comment (InnocentBug) [reply .github/workflows/docker-ci.yml:162]: Yeah if you can do that, that works too.
Isn't there a way to set all seeds and then it is determinstic?
Computers are deterministic machines after all.

Whatever works.

Comment (InnocentBug) [reply .github/workflows/docker-ci.yml:162]: https://github.com/SSAGESLabs/PySAGES/actions/runs/3298840961/jobs/5441644009

Comment (yihengwuKP) [reply .github/workflows/docker-ci.yml:162]: cool I'll do that. I said that because I have a vague memory telling me that openmm might not be totally trustworthy on that. I found similar quote in their documentation: 

> 10.4. Determinism
> Whether a simulation is deterministic will depend on what plaform you run on in addition to what settings/methods you use. For instance, as of this writing, using PME on the Reference, OpenCL, and double-precision CUDA will result in deterministic simulations. Single-precision CUDA and CPU platforms are not deterministic in this case. However, none of this behavior is guaranteed in future versions. In many cases it will still result in an identical trajectory. If determinism is a critical for your needs, you should carefully check to ensure that your settings and platform allow for this.

I am just lazy to deal with that lol. But anyways, I'll add it to the script!

Comment (InnocentBug) [reply .github/workflows/docker-ci.yml:162]: sure, I understand, anyway of checking it is doing the right thing is fine

**Comment by yihengwuKP** [comment]
ah, it seems that I can't put my input PDB into `examples/inputs` because now the gitignore ignores it `example/**/*.pdb`. Any suggestions on quickly fixing this without hacking?

**Comment by yihengwuKP** [comment]
Ah, I see, I'll use `git add -f`

File: pysages/colvars/angles.py
Comment (pabloferz) [inline pysages/colvars/angles.py:206]: What I had in mind was to return `a` and `b` here and compute `q` and `phi` on the `RingAmplitude` and `RingPhaseAngle`, respectively:
```suggestion
    return np.array([a, b])
```

File: pysages/colvars/angles.py
Comment (pabloferz) [inline pysages/colvars/angles.py:242]:
```suggestion
    a, b = ring_puckering_coordinates(rs)
    return np.arctan2(b, a)
```

Comment (yihengwuKP) [reply pysages/colvars/angles.py:242]: I see, but there might be a need for people to get both the q and the phi?

Comment (pabloferz) [reply pysages/colvars/angles.py:242]: That's fair. Not sure what would the most common scenario would be. Do you have any opinions on this @ricalessandri.

Comment (yihengwuKP) [reply pysages/colvars/angles.py:242]: I think the most useful one is the phase angle with the real info, the amplitude, at least in the 5-membered ring in RNA, is usually calculated with the phase angle. I haven't seen people who only calculate the amplitude, because usually, it fluctuates around some eq value, but phase angle can switch between different conformations.

Comment (pabloferz) [reply pysages/colvars/angles.py:242]: Ok. Let's leave as you have it now, and we will revisit if/when there's a demand for something more flexible.

**Review by pabloferz — APPROVED** [review]

File: pysages/colvars/angles.py
Comment (yihengwuKP) [reply pysages/colvars/angles.py:242]: cool, thanks for the review! I'll try to add the test that Ludwig suggested today!

**Comment by yihengwuKP** [comment]
OK, I've added the testing for phase angle like @InnocentBug had suggested.

File: examples/openmm/cv_pucker/phase_angle.txt
Comment (InnocentBug) [reply examples/openmm/cv_pucker/phase_angle.txt:1]: Same with output.dcd, that shouldn't be tracked either

File: examples/openmm/cv_pucker/sugar_coords.txt
Comment (InnocentBug) [inline examples/openmm/cv_pucker/sugar_coords.txt:1]: this is also an output and should be tracked.

File: .github/workflows/docker-ci.yml
Comment (InnocentBug) [inline .github/workflows/docker-ci.yml:162]: you do the work twice here.
you define the SCRIPT variable, but then you also write down the command again.
One is fine.

**Review by InnocentBug — CHANGES_REQUESTED** [review]

**Comment by InnocentBug** [comment]
Thanks for adding the test.
But we need to make sure to not track any output in git repos.
And so far, I don't think you are actually executing the test after the simulation run is done.

File: examples/openmm/cv_pucker/phase_angle.txt
Comment (yihengwuKP) [reply examples/openmm/cv_pucker/phase_angle.txt:1]: ah, good point

File: .github/workflows/docker-ci.yml
Comment (yihengwuKP) [reply .github/workflows/docker-ci.yml:162]: thanks for the code review, I'll fix it

**Comment by yihengwuKP** [comment]
That's true, I tested it locally and got excited and forgot about it, I'll fix everything soon.

**Review by InnocentBug — APPROVED** [review]

**Comment by pabloferz** [comment]
Thanks for this @yihengwuKP!

**Pull request merged**
